File: agent-driven-architecture-main/examples/typescript/src/spine/ports/sink.ts
// ── spine/ports/sink — the perform seam (G9) ────────────────────────────────
// INTERFACES ONLY (C11).
//
// `perform` accepts a KeyedEffect and NOTHING ELSE, so an effect can never
// cross this seam without the idempotency key 14.6 depends on.
//
//   LIVE      perform once, for real
//   REPLAY    collect the descriptor; TOUCH NOTHING
//   RECOVERY  re-drive un-acknowledged effects; the sink dedupes on the key
//
// REPLAY and RECOVERY are constructed and tested, not merely declared — the
// shipped reference declared REPLAY in both ports and constructed it nowhere.

import type { EffectBase } from "../pure/effect";
import type { KeyedEffect } from "../pure/keyed-effect";

export type PerformMode = "LIVE" | "REPLAY" | "RECOVERY";

export interface Sink {
  perform(keyed: KeyedEffect<EffectBase>, mode: PerformMode): void;
}

File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/ports/Sink.kt
// ── spine/ports/sink — the perform seam (G9) ───────────────────────────────
// perform accepts a KeyedEffect and NOTHING ELSE. The key is the idempotency key
// 14.6 depends on, and it is derived from the committed step index, so the fold
// cannot mint it and a caller cannot forget it.

package adr.spine.ports

import adr.spine.pure.KeyedEffect
import adr.spine.pure.PerformMode

interface Sink {
    fun perform(keyed: KeyedEffect, mode: PerformMode)
}
